firmware-nucleo/smart-air-hockey-table/App/driver_led.c

```diff
@@ -62,37 +65,49 @@ void Driver_LED_SetColor(uint_fast8_t col, uint_fast8_t row, Color color)
 		uint_fast8_t r_bit = (color.red >> (COLOR_8BIT_OFFSET - i)) & 0b1;
 		uint_fast8_t g_bit = (color.green >> (COLOR_8BIT_OFFSET - i)) & 0b1;
 		uint_fast8_t b_bit = (color.blue >> (COLOR_8BIT_OFFSET - i)) & 0b1;
-		led_buffer[(r_offset + i) * LED_CHANNELS + channel] = (!r_bit) * COLOR_COMPARE_OFF + r_bit * COLOR_COMPARE_ON;
-		led_buffer[(g_offset + i) * LED_CHANNELS + channel] = (!g_bit) * COLOR_COMPARE_OFF + g_bit * COLOR_COMPARE_ON;
-		led_buffer[(b_offset + i) * LED_CHANNELS + channel] = (!b_bit) * COLOR_COMPARE_OFF + b_bit * COLOR_COMPARE_ON;
+		led_buffer[channel][r_offset + i] = (!r_bit) * COLOR_COMPARE_OFF + r_bit * COLOR_COMPARE_ON;
+		led_buffer[channel][g_offset + i] = (!g_bit) * COLOR_COMPARE_OFF + g_bit * COLOR_COMPARE_ON;
+		led_buffer[channel][b_offset + i] = (!b_bit) * COLOR_COMPARE_OFF + b_bit * COLOR_COMPARE_ON;
 	}
 	is_transfer_requested = true;
 }
 
 void Driver_LED_Clear()
 {
-	for (uint_fast32_t i = 0; i < LED_DATA_LENGTH; i++)
+	for (uint_fast32_t channel = 0; channel < LED_CHANNELS; channel++)
 	{
-		led_buffer[i] = COLOR_COMPARE_OFF;
+		for (uint_fast32_t i = 0; i < LED_DATA_LENGTH / LED_CHANNELS; i++)
+		{
+			led_buffer[channel][i] = COLOR_COMPARE_OFF;
+		}
 	}
 	is_transfer_requested = true;
 }
 
 void Driver_LED_Tick()
 {
-	if (is_transfer_requested && !is_transfer_active)
+	if (is_transfer_requested && transfers_active == 0)
 	{
 		is_transfer_requested = false;
-		is_transfer_active = true;
+		transfers_active = LED_CHANNELS;
 
 		// Configure GPDMA Channel to initiate transfer
-		LL_DMA_ConfigAddresses(GPDMA1, LL_DMA_CHANNEL_0, (uint32_t) led_buffer, (uint32_t) &(TIM2->DMAR));
-		LL_DMA_SetBlkDataLength(GPDMA1, LL_DMA_CHANNEL_0, LED_BUFFER_LENGTH);
+		LL_DMA_ConfigAddresses(GPDMA1, LL_DMA_CHANNEL_0, (uint32_t) led_buffer[0], (uint32_t) &(TIM2->CCR1));
+		LL_DMA_ConfigAddresses(GPDMA1, LL_DMA_CHANNEL_1, (uint32_t) led_buffer[1], (uint32_t) &(TIM2->CCR2));
+		LL_DMA_ConfigAddresses(GPDMA1, LL_DMA_CHANNEL_2, (uint32_t) led_buffer[2], (uint32_t) &(TIM2->CCR3));
+		LL_DMA_ConfigAddresses(GPDMA1, LL_DMA_CHANNEL_3, (uint32_t) led_buffer[3], (uint32_t) &(TIM2->CCR4));
+		LL_DMA_SetBlkDataLength(GPDMA1, LL_DMA_CHANNEL_0, LED_CHANNEL_LENGTH);
+		LL_DMA_SetBlkDataLength(GPDMA1, LL_DMA_CHANNEL_1, LED_CHANNEL_LENGTH);
+		LL_DMA_SetBlkDataLength(GPDMA1, LL_DMA_CHANNEL_2, LED_CHANNEL_LENGTH);
+		LL_DMA_SetBlkDataLength(GPDMA1, LL_DMA_CHANNEL_3, LED_CHANNEL_LENGTH);
 		LL_DMA_EnableChannel(GPDMA1, LL_DMA_CHANNEL_0);
+		LL_DMA_EnableChannel(GPDMA1, LL_DMA_CHANNEL_1);
+		LL_DMA_EnableChannel(GPDMA1, LL_DMA_CHANNEL_2);
+		LL_DMA_EnableChannel(GPDMA1, LL_DMA_CHANNEL_3);
 	}
 }
 
-void GPDMA1_Channel0_Handler()
+void GPDMA1_Channel0123_Handler()
 {
-	is_transfer_active = false;
+	transfers_active--;

```

Comment:
Is the handler called every time a channel is flushed?

Reply:
Are the calls to the handler atomic? Cuz if not we have a data race that will result in errors.